### PR TITLE
made dynamic embed list function in message.service.

### PR DESCRIPTION
**Pull request by CallumCoy**
Also, put it into the cat plugin (currently the only plugin that uses it).

### CODE REVIEW

File: src/services/message.service.ts [hunk 1]
```diff
@@ -77,6 +77,36 @@ export class MessageService {
...
+    const embedItem = new MessageEmbed();
+    embedItem.setColor('#0099ff').setTitle(title);
+
+    // Splits the list into numCols as evenly as possible.
+    const columns = new Array(numCols).fill(0).map((_) => listItems.splice(0, numRows));
+
+    // Cycles through each column inserting them, and also notes the alphabetic range
+    columns.forEach((column) => {
+      const fromLetter = column[0].charAt(0);
+      const toLetter = column[column.length - 1].charAt(0);
```
Comment (c650) [inline src/services/message.service.ts:98]: careful w this

File: src/services/message.service.ts [hunk 2]
```diff
@@ -77,6 +77,36 @@ export class MessageService {
...
+    const maxCol = 3;
+    const maxRows = 10; // This is a soft limit
+
+    // Setting up number of columns and rows for the list
+    const numCols = Math.min(maxCol, Math.ceil(listItems.length / maxRows));
+    const numRows = Math.ceil(listItems.length / numCols);
+
+    // Sets up the embed
+    const embedItem = new MessageEmbed();
+    embedItem.setColor('#0099ff').setTitle(title);
```
Comment (tanndlin) [inline src/services/message.service.ts:90]: Dont set the title in this function to make it simpler. bc each embed will need a different title, its best to set it in the plugin

Comment (c650) [reply src/services/message.service.ts:90]: perhaps color also configurable. maybe make options Object w optional params

Comment (CallumCoy) [reply src/services/message.service.ts:90]: Easier to let the plugin do it all @c650 when the messageEmbed is returned (I tried to send in a optional messageEmbed but the foreach didn't like it).  So I'll just take out the title like tanndlin said and the plugin can deal with it.  Just thought it was common enough to stay in it.

Comment (c650) [reply src/services/message.service.ts:90]: gotcha.

**Review by c650 — APPROVED** [review]

File: src/services/message.service.ts (hunk 2 above)
Comment (CallumCoy) [reply src/services/message.service.ts]: In the end, I made it a dynamic choice.

File: src/app/plugins/cat.plugin.ts [hunk 3]
```diff
@@ -81,6 +77,16 @@ export class CatPlugin extends Plugin {
       .catch((err) => this.container.loggerService.warn(err));
   }
 
+  private async _getListEmbed() {
+    const breedsAsArray = this._breeds.map((breedData: { name: string; id: string }) => {
+      return breedData.name;
+    });
+    this._embedBreeds = await this.container.messageService.generateEmbedList(
+      breedsAsArray,
+      new MessageEmbed()
```
Comment (tanndlin) [inline src/app/plugins/cat.plugin.ts:86]: What's the point in passing in `new MessageEmbed()`?

File: src/services/message.service.ts [hunk 4]
```diff
@@ -77,25 +77,30 @@ export class MessageService {
...
-    const embedItem = new MessageEmbed();
-    embedItem.setColor('#0099ff').setTitle(title);
+    let embedItem: MessageEmbed;
+
+    if (typeof arg === 'object') {
+      embedItem = arg;
+    } else {
+      embedItem = new MessageEmbed();
+      embedItem.setColor('#0099ff').setTitle(arg);
+    }
```
Comment (tanndlin) [inline src/services/message.service.ts:95]: What's this?

File: src/app/plugins/cat.plugin.ts [hunk 5]
```diff
@@ -18,7 +18,7 @@ export class CatPlugin extends Plugin {
 
   private _API_URL: string = 'https://api.thecatapi.com/v1/';
   private _breeds: Breed[] = [];
-  private _embedBreeds = new MessageEmbed();
+  private _embedBreeds: Maybe<MessageEmbed>;
```
[the remark this answers is not in the document]
Comment (CallumCoy) [reply src/app/plugins/cat.plugin.ts:21]: it's so, I can check if it needs to be generated or not later on in the code.

File: src/app/plugins/cat.plugin.ts (hunk 3 above)
Comment (CallumCoy) [reply src/app/plugins/cat.plugin.ts:86]: oops mb that was from testing

File: src/app/plugins/cat.plugin.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (tanndlin) [reply src/app/plugins/cat.plugin.ts:21]: oh, my mistake

File: src/services/message.service.ts [hunk 6]
```diff
@@ -77,25 +77,30 @@ export class MessageService {
...
     // Setting up number of columns and rows for the list
     const numCols = Math.min(maxCol, Math.ceil(listItems.length / maxRows));
     const numRows = Math.ceil(listItems.length / numCols);
 
-    // Sets up the embed
-    const embedItem = new MessageEmbed();
-    embedItem.setColor('#0099ff').setTitle(title);
+    let embedItem: MessageEmbed;
+
+    if (typeof arg === 'object') {
```
Comment (c650) [inline src/services/message.service.ts:90]: i dont like this

Comment (CallumCoy) [reply src/services/message.service.ts:90]: the idea in general or the typeof command?

**Comment by tanndlin** [comment]
Having the letters at the top assumes the list is sorted, but its not necessarily sorted, i feel that should be an option

File: src/services/message.service.ts (hunk 6 above)
Comment (tanndlin) [reply src/services/message.service.ts:91]: typeof, make an options object as optional not a multi type arg

File: src/app/plugins/cat.plugin.ts [hunk 7]
```diff
@@ -50,7 +45,10 @@ export class CatPlugin extends Plugin {
 
     if (args[0].includes('breed')) {
       //Simply return the list of supported breeds
-      message.reply(this._embedBreeds);
+      if (!this._embedBreeds) {
+        await this._getListEmbed();
+      }
+      message.reply(this._embedBreeds || 'Failed to load breeds.');
```
Comment (c650) [inline src/app/plugins/cat.plugin.ts:51]: nit: do the `if` check in the `_getListEmbed` func so that the caller doesnt have to think about when to call it

**Review by c650 — APPROVED** [review]

File: src/services/message.service.ts [hunk 8]
```diff
@@ -77,6 +77,51 @@ export class MessageService {
...
+
+    // Title was sent inserts it into the embed.
+    if (options.title) {
+      embedItem.setColor('#0099ff').setTitle(options.title);
+    }
+
+    // finds out if the list is sorted or not
+    const temp = listItems;
+    const sortedList = !!temp.reduce(
+      (prevRes: any, item: any) => prevRes !== false && item >= prevRes && item
```
Comment (tanndlin) [inline src/services/message.service.ts:99]: Interesting

File: src/app/plugins/cat.plugin.ts [hunk 9]
```diff
@@ -45,9 +45,7 @@ export class CatPlugin extends Plugin {
 
     if (args[0].includes('breed')) {
       //Simply return the list of supported breeds
-      if (!this._embedBreeds) {
-        await this._getListEmbed();
-      }
+      await this._getListEmbed();
```
Comment (tanndlin) [inline src/app/plugins/cat.plugin.ts:48]: I think corgo said this on the last pr, this should return the list, and just stick it in the `message.reply`

File: src/app/plugins/cat.plugin.ts [hunk 10]
```diff
@@ -45,9 +45,7 @@ export class CatPlugin extends Plugin {
...
       message.reply(this._embedBreeds || 'Failed to load breeds.');
```
Comment (tanndlin) [inline src/app/plugins/cat.plugin.ts:49]: `await`

**Review by tanndlin — COMMENTED** [review]
other than those, lgtm

File: src/app/plugins/cat.plugin.ts (hunk 10 above)
Comment (CallumCoy) [reply src/app/plugins/cat.plugin.ts:49]: what do you mean await?

Comment (tanndlin) [reply src/app/plugins/cat.plugin.ts:49]: await the message reply

**Pull request merged**